### PR TITLE
Query parsing with whitespace and multifield support. ARXIVNG-1897

### PR DESCRIPTION
**Pull request by JaimieMurdock**
Okay, lots of PRs this evening.

This review will be merged into the branch for ARXIVNG-1893. I'd suggest taking a look at that PR first, which just sets up the scaffolding for this branch: https://github.com/arXiv/arxiv-search/pull/232

This PR includes a `/query` endpoint that matches the documentation at https://arxiv.org/help/api/user-manual. However, there are a few things missing:
- *boolean support* - currently, all fields default to `AND`. Bad things will happen if you insert the booleans in your `search_query` query parameter. (Hopefully a 400 error complaining about the syntax!)
- *parentheses support* - there just isn't any. We don't have the mechanisms in place for nested `domain.advanced.FieldedSearchList` objects.

For testing use the following command in conjunction with an elasticsearch environment (`docker-compose up`) that has been initialized with the instructions in the `README.md`:

`FLASK_APP=classic_api.py JWT_SECRET=foosecret FLASK_DEBUG=1 ELASTICSEARCH_HOST=127.0.0.1 pipenv run flask run`

Here are some sample queries to look at:
- http://127.0.0.1:5000/query?search_query=all:%22physics+universes%22+ti:massless (should show one result)
- http://127.0.0.1:5000/query?search_query=all:none&id_list=1710.08254v1 (expect one result)
- http://127.0.0.1:5000/query?search_query=all:none&id_list=1710.08254 (expect same result as above)
- http://127.0.0.1:5000/query?search_query=all:none (expect 3 results)
- http://127.0.0.1:5000/query?id_list=1710.08254v1,1610.07526v3

## Known issues
*NO unit tests* - This PR does not include unit tests. I'll need to go back and add them, but am still settling on a representation for the parsed term+boolean representation that would make the tests fail based on design decisions - not based on functionality, which is tested by the sample queries above.

*versioned identifiers not working with id_list param*
The following two URLs should have identical results, but do not.
- http://127.0.0.1:5000/query?id_list=1710.08254,1610.07526
- http://127.0.0.1:5000/query?id_list=1710.08254v1,1610.07526v3

This is due to the implementation of the `get_document` function in the search API and is likely a bug in the current deployment. 

Note that if a `search_query` parameter is present the bug disappears - this is because all results are returned from the search API, and then are filtered in the controller method.

### CODE REVIEW

File: search/controllers/api/__init__.py [hunk 1]
```diff
@@ -108,8 +119,59 @@ def classic_query(params: MultiDict) -> Tuple[Dict[str, Any], int, Dict[str, Any
...
+    if not id_list and not raw_query:
+        raise BadRequest("Either a search_query or id_list must be specified for the classic API.")
+
+    if raw_query:
+        # migrate search_query -> query variable
+        params['query'] = raw_query
+        del params['search_query']
+
+        # pass to normal search, which will handle parsing
+        data, _, _ = search(params)
```
Comment (erickpeirson) [inline search/controllers/api/__init__.py]: Beware passing through unsanitized client data.

File: search/controllers/api/__init__.py [hunk 2]
```diff
@@ -108,8 +119,59 @@ def classic_query(params: MultiDict) -> Tuple[Dict[str, Any], int, Dict[str, Any
...
+    
+    if id_list and not raw_query:
+        # Process only id_lists.
+        # Note lack of error handling to implicitly propogate any errors. 
+        # Classic API also errors if even one ID is malformed.
```
Comment (erickpeirson) [inline search/controllers/api/__init__.py]: That's probably a feature. Better to complain loudly than proceed with unannounced errors

File: search/controllers/api/__init__.py [hunk 3]
```diff
@@ -206,7 +268,7 @@ def _to_classification(value: str, query_terms: List) \
     elif value in taxonomy.definitions.ARCHIVES:
         klass = taxonomy.Archive
         field = 'archive'
-    elif value in taxonomy.definitions.CATEGORIES:
+    elif value in taxonomy.definitionss.CATEGORIES:
```
Comment (erickpeirson) [inline search/controllers/api/__init__.py]: looks like a typo (``definitionss``)

Comment (mhl10) [inline search/controllers/api/__init__.py]: typo?

Comment (JaimieMurdock) [reply search/controllers/api/__init__.py]: This  will be corrected by PR #234

Comment (JaimieMurdock) [reply search/controllers/api/__init__.py]: will be corrected by PR #234

**Pull request merged**